File: Exercism/javascript/perfect-numbers/perfect-numbers.js
'use strict'

const numbers = function * (till) {
  for (let i = 1; i < till + 1; i += 1) {
    yield i
  }
}

const isDividerOf = (a) => (b) => a % b === 0

const aliquotSum = (isDivider) => (int) =>
  [...numbers(parseInt(int / 2))]
  .filter(isDivider)
  .reduce((a, b) => a + b, 0)

const perfect = (aliquot, int) => aliquot === int
const abundant = (aliquot, int) => aliquot > int
const deficient = (aliquot, int) => aliquot < int

const matchType = (int) => (aliSum) => (fn) => fn(aliSum(int), int)

const fnName = (__, fn) => fn.name

const combine = (...fns) => (int) => fns.reduce((int, fn) => fn(int), int)

export const classify = (int) => {
  if(!Number.isInteger(int) || int <= 0) {
    throw new Error('Classification is only possible for natural numbers.')
  }
  return [perfect, abundant, deficient]
    .filter(
      combine(
        isDividerOf,
        aliquotSum,
        matchType(int)
      )(int)
    ).reduce(fnName, '')
}

File: Exercism/python/perfect-numbers/perfect_numbers.py
def deviders(number: int) -> int:
    """Generates all deviders for a given number"""
    for i in range(1, number // 2 + 1):
        if number % i == 0:
            yield i


def classify(number: int) -> str:
    """Determine the class of a number"""
    if number <= 0:
        raise ValueError("Error: positive integer required")
    aliquot_sum = sum(deviders(number))
    if aliquot_sum == number:
        return "perfect"
    elif aliquot_sum > number:
        return "abundant"
    elif aliquot_sum < number:
        return "deficient"
